### PR TITLE
342305: Automate the Deploy Keys in Github flux services repo for each environment

### PR DESCRIPTION
**Pull request by ajaysoni-cap**
# **What this PR does / why we need it**:
Automate the Deploy Keys in Github flux services repo for each environment
Relates to ADO Work Item [AB#342305](https://dev.azure.com/defragovuk/93c65135-d16b-49d6-a191-4a5313532779/_workitems/edit/342305)

# **Special notes for your reviewer**
N/A

# Testing
N/A

# **How does this PR make you feel**:
gif

### CODE REVIEW

File: infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1 [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+.PARAMETER AppKeySecretName
+    Mandatory. Github App Private Key Secret Name
+.PARAMETER Environment
+    Mandatory. Application Environment for Key Title
+.PARAMETER SSHPublicKeySecretName
+    Mandatory. Key Vault secret name for the SSH key
+.PARAMETER PSHelperDirectory
+    Mandatory. Directory Path of PSHelper module
+.EXAMPLE
+    .\Add-SshKeyToGitHubRepo.ps1 -AppId <AppId> -AppKey <AppKey> -Environment <Environment> -SSHPublicKeySecretName <SSHPublicKeySecretName> -PSHelperDirectory <PSHelperDirectory>
```
Comment (ppotturi) [inline infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1:17]: I think the parameters names should be AppIdSecretName and AppKeySecretName.

File: infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1 [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        $repoKeysUrl = "https://api.github.com/repos/defra/adp-flux-services/keys"
+        
+        Write-Debug "Reading all the deploy keys..."
+        [array]$keys = Invoke-RestMethod -Method Get -Uri $repoKeysUrl -Headers $headers
+        if ($keys) {
+            Write-Debug "Check if key already exists..."
+            [object]$existingKey = $keys | Where-Object { $_.title -eq $keyTitle }
+
+            if ($existingKey -and $existingKey -ne '') {
+                Write-Output "Deleting existing key '$keyTitle' having id '$($existingKey.id)'."
```
Comment (ppotturi) [inline infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1:141]: Write-host?

File: infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1 [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+                Invoke-RestMethod -Method Delete -Uri ("$repoKeysUrl/{0}" -f $existingKey.id) -Headers $headers | Out-Null
+            }
+        }
+
+        $body = @{
+            "title" = $keyTitle
+            "key" = $DeployKey
+            "read_only" = $false
+        } | ConvertTo-Json
+        Write-Output "Adding new key '$keyTitle'..."
```
Comment (ppotturi) [inline infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1:151]: Write-Host?

File: infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1 [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    $appId = Get-AzKeyVaultSecret -VaultName $KeyVaultName -Name $AppIdSecretName -AsPlainText
+    $appKey = Get-AzKeyVaultSecret -VaultName $KeyVaultName -Name $AppKeySecretName -AsPlainText
+    $jwt = Get-GithubJwt -AppId $appId -AppKey $appKey
+
+    $installationToken = Get-InstallationToken -GitHubJwtToken $jwt
+
+    $deployKey = Get-AzKeyVaultSecret -VaultName $KeyVaultName -Name $SSHPublicKeySecretName -AsPlainText
+    Set-NewDeployKey -InstallationToken $installationToken -Environment $Environment -DeployKey $deployKey
+
+    $exitCode = 0
```
Comment (ppotturi) [inline infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1:196]: Add more logging here?

File: infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1 [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    }
+    process {
+        $headers = @{
+            "Authorization"        = "Token $InstallationToken"
+            "Accept"               = "application/vnd.github+json"
+            "X-GitHub-Api-Version" = "2022-11-28"
+        }
+    
+        $keyTitle = "$($Environment.ToLower())_01"
+        $repoKeysUrl = "https://api.github.com/repos/defra/adp-flux-services/keys"
```
[the remark this answers is not in the document]
Comment (DaRoszko) [reply infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1:132]: would agree, also be nice to work in sandbox etc

File: infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1 [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+Write-Debug "${functionName}:AppIdSecretName=$AppIdSecretName"
+Write-Debug "${functionName}:AppKeySecretName=$AppKeySecretName"
+Write-Debug "${functionName}:Environment=$Environment"
+Write-Debug "${functionName}:SSHPublicKeySecretName=$SSHPublicKeySecretName"
+Write-Debug "${functionName}:KeyVaultName=$KeyVaultName"
+
+try {
+    Import-Module $PSHelperDirectory -Force
+
+    $appId = Get-AzKeyVaultSecret -VaultName $KeyVaultName -Name $AppIdSecretName -AsPlainText
```
Comment (DaRoszko) [inline infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1:187]: We can't do this - remember the debug logs? can we use the task

**Review by DaRoszko — CHANGES_REQUESTED** [review]

File: infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1 (hunk 2 above)
Comment (ajaysoni-cap) [reply infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1]: Fixed.

File: infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1 (hunk 3 above)
Comment (ajaysoni-cap) [reply infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1]: Fixed.

File: infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1 (hunk 6 above)
Comment (ajaysoni-cap) [reply infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1]: Implementation changed.

File: infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1 (hunk 4 above)
Comment (ajaysoni-cap) [reply infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1:204]: Changed to CLI call with PowerShell helpers that will add additional debug logging..

File: infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1 (hunk 5 above)
[the remark this answers is not in the document]
Comment (ajaysoni-cap) [reply infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1]: Parameterised the org and repository with defaults in the definition.

File: infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1 (hunk 1 above)
Comment (ajaysoni-cap) [reply infra/core/env/scripts/Add-SshKeyToGitHubRepo.ps1]: Fixed.

**Review by DaRoszko — APPROVED** [review]

**Review by kbabigumira — APPROVED** [review]

**Pull request merged**